### ISSUE TITLE
AP_duration_half_width rounded of to nearest 0.1 ms?

### ISSUE THREAD

**Issue by robban80** [BlueBrain/eFEL#142]
Hi,
I'm trying out the library so possible that I have done something wrong, but seems like spike durations are rounded of to closest 0.1 ms (have run ~200 traces that all belongs to discrete values 0.9, 1.0 or 1.1). Is this the intended behavior?

Here's the function I use:

```
def extract_features(t,y):
    # extracts electrical features from trace using bbp efel library
    
    import efel
    
    # code based on example example code from: https://github.com/BlueBrain/eFEL --------------------
    
    # A 'trace' is a dictionary
    trace1 = {}

    # Set the 'T' (=time) key of the trace
    trace1['T'] = t

    # Set the 'V' (=voltage) key of the trace
    trace1['V'] = y

    # Set the 'stim_start' (time at which a stimulus starts, in ms)
    # key of the trace
    # Warning: this need to be a list (with one element)
    trace1['stim_start'] = [100]

    # Set the 'stim_end' (time at which a stimulus end) key of the trace
    # Warning: this need to be a list (with one element)
    trace1['stim_end'] = [1000]

    # Multiple traces can be passed to the eFEL at the same time, so the
    # argument should be a list
    traces = [trace1]

    # Now we pass 'traces' to the efel and ask it to calculate the feature
    # values
    traces_results = efel.getFeatureValues(traces,
                                           ['AP_begin_voltage', \
                                            'AP_amplitude', \
                                            'AP_duration_half_width', \
                                            'voltage_base']
                                          )
```

**Comment by wvangeit** [comment]
Yes, that could make sense.
The default interpolation frequency of eFEL is 0.1 ms. You can change it by using:
```
efel.setDoubleSetting('interp_step', 0.1)
```

**Comment by robban80** [comment]
Thank you, but when I change this parameter it seems like other features are less well approximated. Specifically it seems like AP_begin_voltage, that I have used as an approximation for threshold, is changed so that for some traces the value reported is the same as the voltage_base (about -83).

I also found the feature spike_half_width that seems to be continuous, so maybe I can simply use this feature instead?

So to sum up my two questions:
1. Can I use AP_begin_voltage to find the threshold of the spike (or what should I use)?
2. What's the difference between the AP_duration_half_width and spike_half_width?

Thanks

**Comment by wvangeit** [comment]
I think normally AP_begin_voltage is the right feature to use. Could you share the trace you have problems with in some way ? Then I could have a look.

About AP_duration_half_width and spike_half_width. There are more or less similar, but developed at different times. Will have a look at it in more detail, will probably immediately add some descriptions to the docs.

**Comment by robban80** [comment]
Thank you, that would be useful!
[VOLTAGE.pkl.zip](https://github.com/BlueBrain/eFEL/files/2434064/VOLTAGE.pkl.zip)

have added the traces to this tread in a zip file. Inside is a pickled dictionary with keys 'tm' and 'vm', where 'vm' holds all the traces and can be opened by something like:

```
import pickle
import matplotlib.pyplot as plt

def load_obj(name ):
    with open(name, 'rb') as f:
        return pickle.load(f) 

data = load_obj('VOLTAGE.pkl')

for trace in data['vm']:
    # do something to data e.g. plotting
    plt.plot(data['tm'], data['vm'][trace]
plt.show
```

**Comment by wvangeit** [comment]
Had some time to look at this.
What seems to be happening is the following.
The AP_begin_voltages features goes through the spike and checks where the crosses a certain threshold. This has happen in a certain window. This window is currently set to 3 sampling points. Obviously this window will be affected by the interpolation frequency ...
I see two options to solve this:
- First, in the long term this should be fixed by making this window time-independent. Since it will affect all the users, and older results, it's probably best to make a newer feature for this.
- In the short term, I could solve it for you by making the window size a user-configurable option. Do you mind if I put one of your traces in the repo, to use it as part of the unit testing ?

**Comment by robban80** [comment]
Ok thank you, sure let's do that. 

I don't mind if the traces are used for testing purpose, so please go ahead and add one trace to the repo.

**Comment by wvangeit** [comment]
Ok, implemented this in https://github.com/BlueBrain/eFEL/pull/143
If you do a 
```
pip install pip --upgrade
```
you should get the latest version.

While adding the test, I realized the issue is slightly more complicated. There is an effect of the window, but the core problem is actually that in your trace the derivate threshold (use to detect the beginning of the spike) is crossed in the steep part right after the beginning of the stimulus. The size of the window determines if this is detected or not.
So you could actually also solve your problem by increasing the threshold a bit:
```
efel.setDoubleSetting('DerivativeThreshold', 30.0) 
```
(the default value is 20.0 mV/ms)

Or you can control the window after you set the interp_step to e.g. 30 steps instead of the current 3:
```
efel.setIntSetting('DerivativeWindow', 30)       
```

Hope that helps
